eth/eventdatahandler/event_data_handler.go

```diff
@@ -0,0 +1,536 @@
+	shareEncryptionKeyProvider ShareEncryptionKeyProvider
+	keyManager                 types.KeyManager
+	beacon                     beaconprotocol.BeaconNode
+	storageMap                 *qbftstorage.QBFTStores
+	fullNode                   bool
+	logger                     *zap.Logger
+	metrics                    metrics
+}
+
+func New(
+	eventDB eventDB,
+	client *executionclient.ExecutionClient,
+	taskExecutor taskExecutor,
+	operatorData *storage.OperatorData,
+	shareEncryptionKeyProvider ShareEncryptionKeyProvider,
+	keyManager types.KeyManager,
+	beacon beaconprotocol.BeaconNode,
+	storageMap *qbftstorage.QBFTStores,
+	opts ...Option,
+) (*EventDataHandler, error) {
+	abi, err := contract.ContractMetaData.GetAbi()
+	if err != nil {
+		return nil, fmt.Errorf("get contract ABI: %w", err)
+	}
+
+	filterer, err := client.Filterer()
+	if err != nil {
+		return nil, fmt.Errorf("create contract filterer: %w", err)
+	}
+
+	edh := &EventDataHandler{
+		eventDB:                    eventDB,
+		taskExecutor:               taskExecutor,
+		abi:                        abi,
+		filterer:                   filterer,
+		operatorData:               operatorData,
+		shareEncryptionKeyProvider: shareEncryptionKeyProvider,
+		keyManager:                 keyManager,
+		beacon:                     beacon,
+		storageMap:                 storageMap,
+		logger:                     zap.NewNop(),
+		metrics:                    nopMetrics{},
+		shareMap:                   sharemap.New(),
+	}
+
+	for _, opt := range opts {
+		opt(edh)
+	}
+
+	return edh, nil
+}
+
+func (edh *EventDataHandler) HandleBlockEventsStream(blockEventsCh <-chan eventbatcher.BlockEvents, executeTasks bool) (uint64, error) {
+	var lastProcessedBlock uint64
+
+	for blockEvents := range blockEventsCh {
+		logger := edh.logger.With(fields.BlockNumber(blockEvents.BlockNumber))
+
+		logger.Info("processing block events")
+		tasks, err := edh.processBlockEvents(blockEvents)
+		if err != nil {
+			return 0, fmt.Errorf("process block events: %w", err)
+		}
+
+		lastProcessedBlock = blockEvents.BlockNumber
+
+		logger.Info("processed block events")
+
+		if !executeTasks || len(tasks) == 0 {
+			continue
+		}
+
+		logger.Info("executing tasks", fields.Count(len(tasks)))
+
+		tasks = edh.filterSupersedingTasks(tasks)
+		for _, task := range tasks {
+			logger = logger.With(fields.Type(task))
+			logger.Debug("going to execute task")
+			if err := task.Execute(); err != nil {
+				// TODO: We log failed task until we discuss how we want to handle this case. We likely need to crash the node in this case.
+				logger.Error("failed to execute task", zap.Error(err))
+			} else {
+				logger.Debug("executed task")
+			}
+		}
+
+		logger.Info("task execution finished", fields.Count(len(tasks)))
+	}
+
+	return lastProcessedBlock, nil
+}
+
+func (edh *EventDataHandler) processBlockEvents(blockEvents eventbatcher.BlockEvents) ([]Task, error) {
+	txn := edh.eventDB.RWTxn()
+	defer txn.Discard()
```

Comment:
should `Discard` be called after `Commit` as well ?

Reply:
I doesn't have to be but can be. 

```go
// Discard discards a created transaction. This method is very important and must be called. Commit
// method calls this internally, however, calling this multiple times doesn't cause any issues. So,
// this can safely be called via a defer right when transaction is created.
```

I preferred to keep `Discard` because it's a must for every transaction and I wanted to explicitly show that transaction would be finalized correctly. I think `Commit` calling `Discard` is implementation detail, so if we remove `Discard`, the one who reads the code would need to check docs to make sure that `Commit` is enough.

Reply:
yeah I understand and agree lets keep it.